### PR TITLE
Add cpm5d aggregation function

### LINKED ISSUE #10419: [Feature] Add a new cpm aggregation function that does not lose precision when downsampled
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

For services with a low rate of requests, for instance 1 per hour, the `cpm` aggregation function loses the requests when it is downsampled to hours or days treating it as effectively 0 requests.
This lose of precision can misleading while being unable to different if there really has been a request or not.

### Use case

Create a new function, for instance `cpm5d` (5 decimals), that doesn't lose precision after downsampling.
We need at least 5 decimals if we want to represent as precise as possible 1 request in a concrete hour of a day, downsampled to days.
The following example is 1 request in a concrete hour or a day. Which means at the end 1 request per day.
1 decimal: 1 / 1440 = 0 // We cannot different anymore
4 decimals: (1 * 10000) / 1440 = 6
5 decimals: (1 * 100000) / 1440 = 69

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by toffentoffen**
`cpm5d` function allows does not lose precision when it is downsampled to hours or days.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #10419.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
@Fine0830 Could you add `Precision is 5` for this? We need to format `cpm5d` through that.

@toffentoffen We need to update CPM configuration to `CPM5D` in OAL, and settings in general and mesh dashboards, after @Fine0830 's work. Also, please fix your UT.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetrics.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.skywalking.oap.server.core.analysis.metrics;
+
+import org.apache.skywalking.oap.server.core.analysis.metrics.annotation.MetricsFunction;
+
+@MetricsFunction(functionName = "cpm5d")
+public abstract class CPM5DecimalsMetrics extends CPMMetrics implements LongValueHolder {
+
+    @Override
+    public void calculate() {
+        this.setValue((this.getTotal() * 100000) / getDurationInMinute());
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetrics.java:28]:
```suggestion
        this.setValue((this.getTotal() * 100_000) / getDurationInMinute());
```

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetricsTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+package org.apache.skywalking.oap.server.core.analysis.metrics;
+
+import org.apache.skywalking.oap.server.core.remote.grpc.proto.RemoteData;
+import org.apache.skywalking.oap.server.core.storage.StorageID;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class CPM5DecimalsMetricsTest {
+
+    public static final long TIME_BUCKET = 2015_11_07_1130L;
```
Comment (toffentoffen) [inline oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetricsTest.java:28]: Does this date look familiar to anyone?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetrics.java (hunk 1 above)
Comment (toffentoffen) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/CPM5DecimalsMetrics.java]: Applied. Thanks for the suggestion.

**Comment by wu-sheng** [comment]
https://github.com/apache/skywalking-booster-ui/pull/239 has merged. Please update the relative dashboards and OAL to adopt this cpm5d function.

**Comment by toffentoffen** [comment]
> [apache/skywalking-booster-ui#239](https://github.com/apache/skywalking-booster-ui/pull/239) has merged. Please update the relative dashboards and OAL to adopt this cpm5d function.

Let's say I have updated all the needed OAL expression to use `cpm5d()` instead of `cmp()`:
For instance `endpoint_cpm` would be:
```
# FROM
# endpoint_cpm = from(Endpoint.*).cpm();
# TO
 endpoint_cpm = from(Endpoint.*).cpm5d(); # easy peasy change XD
```

So now I am trying to update the dashboards. I guess that what I need to do is to change the `calculation`(of the `metricConfig`) of any component in a dashboard that relies on any `*_cpm` metric to use the  `cpm5d`, but I have some doubts about the dashboards. 
Looking at the UI PR, it looks like `cmp5d` is a new calculation https://github.com/apache/skywalking-booster-ui/blob/5cc913a33233d7fb0db151af4e4e444eb3a9a24b/src/hooks/data.ts#L41. 

However, now I am wondering as `cmp5d` is a Calculation, how should I change the following component:
https://github.com/apache/skywalking/blob/1335a48f1c034abc1fe24f6197ee7acfa3118bf0/oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-service.json#L715-L717
This concrete component is the following endpoints list:
endpoints_list
The first column, is configured as "read all values of the duration"(`readMetricsValues`), and then all the values get averaged. 
But here comes the problem, if I change it to `cpm5d`, the list will be broken.
The think is that we still need calculate the average, but then format it differently. 
So we maybe need to have an `Average5d` calculation, and maybe other `[calculation]5d` calculations, or we have developed the solution in a way that doesn't fit correctly.

**Comment by wu-sheng** [comment]
> But here comes the problem, if I change it to cpm5d, the list will be broken.

Could you check the GraphQL to see what happens? Basically, changing the metric name(keep calculation unchanged) should not break anything.
@Fine0830 Anything different here?

**Comment by kezhenxu94** [comment]
> > But here comes the problem, if I change it to cpm5d, the list will be broken.
> 
> Could you check the GraphQL to see what happens? Basically, changing the metric name(keep calculation unchanged) should not break anything. @Fine0830 Anything different here?

By "broken", I think @toffentoffen means the values are not correct, because currently a calculation "average" is already applied to each cell of the first column, if we choose to `cpm5d`, then the cpm for each service is a list (`readMetricsValues` returns a list).

**Comment by wu-sheng** [comment]
I am not following, the read through GraphQL is reading values, and calculated by UI function.
So, the OAL should not be irrelevant.

**Comment by wu-sheng** [comment]
Do you mean we need a new UI function?

**Comment by wu-sheng** [comment]
@Fine0830 I think we need `cpm5d + AvgPreview` for table list calculation.

**Comment by Fine0830** [comment]
> @Fine0830 I think we need `cpm5d + AvgPreview` for table list calculation.

https://github.com/apache/skywalking/pull/10456 has merged. Please feel free to use `cpm5d + AvgPreview` calculation. @toffentoffen

**Comment by toffentoffen** [comment]
Thank you very much @Fine0830 I think with the new `cmd5dAvg` it will work.
Also, thank you @wu-sheng and @kezhenxu94 for helping here.

**Comment by wu-sheng** [comment]
Marc, have you deployed showcase and verify these pages?

**Comment by toffentoffen** [comment]
Okay I think I understood it now.
So in the general service table for instance:
service_list

For each service and metrics, we fetch all the values because we need them to display the spline graph as well as the table value. But the table value is computed on the client side by averaging all the fetch data points for a given service and metric.
That is the main driving force to have `cpm5dAvg` calculation functions on the UI.

**Comment by toffentoffen** [comment]
> Marc, have you deployed showcase and verify these pages?

I wanted to try it out. Will do it next, will let you know @wu-sheng 
Need to figure out how to change the showcase project and point it to my branch.

**Comment by wu-sheng** [comment]
> > Marc, have you deployed showcase and verify these pages?
> 
> I wanted to try it out. Will do it next, will let you know @wu-sheng Need to figure out how to change the showcase project and point it to my branch.

You just need some images. Everything is here, https://github.com/apache/skywalking-showcase/blob/main/Makefile.in

**Comment by wu-sheng** [comment]
@wankai123 Could you re-check this?

**Review by wu-sheng — APPROVED** [review]
LGTM. I noticed there is no `cpm` function doc in the OAL doc. I will update after this gets merged for `cpm` and `cpm5d`.

**Comment by kezhenxu94** [comment]
I have a doubt, now we simply replace the original `cpm` to `cpm5d`, and keep the original metrics name, when users upgrade the OAP but not UI (yet), they will find the related metrics get 100000 times by the real value, this can bring confusion. Also, this is not considered a breaking change, and the metrics data before upgrading and after upgrading will mess up, some of them are original `cpm` and some of them are `x 100000`, the final data make no sense

**Review by kezhenxu94 — CHANGES_REQUESTED** [review]
The doubt should be addressed, thanks.

**Comment by wu-sheng** [comment]
> I have a doubt, now we simply replace the original `cpm` to `cpm5d`, and keep the original metrics name, when users upgrade the OAP but not UI (yet), they will find the related metrics get 100000 times by the real value, this can bring confusion. Also, this is not considered a breaking change, and the metrics data before upgrading and after upgrading will mess up, some of them are original `cpm` and some of them are `x 100000`, the final data make no sense

What do you suggest? Change the metric name in the OAL?

**Comment by wu-sheng** [comment]
BTW, if we change this, our Grafana setup has to follow.

File: oap-server/server-starter/src/main/resources/oal/core.oal [hunk 3]
```diff
@@ -19,15 +19,15 @@
 // Service scope metrics
 service_resp_time = from(Service.latency).longAvg();
 service_sla = from(Service.*).percent(status == true);
-service_cpm = from(Service.*).cpm();
+service_cpm = from(Service.*).cpm5d();
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/oal/core.oal:22]:
```suggestion
service_cpm5d = from(Service.*).cpm5d();
```

Let's follow this to update settings. @wankai123 Once this is merged, Grafana setup should be updated according on showcase.

**Comment by wu-sheng** [comment]
@kezhenxu94 @toffentoffen Another warning, this change(update metric name) could break existing e2e.

**Comment by wu-sheng** [comment]
About the metric name change, `service_cpm` would be about at least cause 66 places changed, including docs and UI setup.

image

**Comment by wu-sheng** [comment]
Close for now. As this is no update for a while, and this will break many metrics from an integration perspective. Let's see whether we really need this.

**Pull request closed without merging**